pallets/nodle-staking/src/lib.rs

```diff
@@ -0,0 +1,2311 @@
+    }
+
+    impl<AccountId: Ord, Balance: Ord> Ord for ValidatorSnapshot<AccountId, Balance> {
+        fn cmp(&self, other: &Self) -> Ordering {
+            self.total.cmp(&other.total)
+        }
+    }
+
+    impl<AccountId: Ord, Balance: Ord> PartialOrd for ValidatorSnapshot<AccountId, Balance> {
+        fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+            Some(self.cmp(other))
+            // Some(self.total.cmp(&other.total))
+        }
+    }
+
+    #[derive(Encode, Decode, Clone, RuntimeDebug)]
+    pub struct Nominator<AccountId, Balance> {
+        pub nominations: OrderedSet<Bond<AccountId, Balance>>,
+        pub total: Balance,
+        pub active_bond: Balance,
+        pub unlocking: Vec<UnlockChunk<Balance>>,
+    }
+
+    impl<
+            AccountId: Ord + Clone,
+            Balance: Copy
+                + sp_std::ops::AddAssign
+                + sp_std::ops::Add<Output = Balance>
+                + sp_std::ops::SubAssign
+                + PartialOrd,
+        > Nominator<AccountId, Balance>
+    {
+        pub fn new(validator: AccountId, amount: Balance) -> Self {
+            Nominator {
+                nominations: OrderedSet::from(vec![Bond {
+                    owner: validator,
+                    amount,
+                }]),
+                total: amount,
+                active_bond: amount,
+                unlocking: Vec::new(),
+            }
+        }
+        pub fn add_nomination(&mut self, bond: Bond<AccountId, Balance>) -> bool {
+            let amt = bond.amount;
+            if self.nominations.insert(bond) {
+                self.total += amt;
+                self.active_bond += amt;
+                true
+            } else {
+                false
+            }
+        }
+        // Returns Some(remaining balance), must be more than MinNominatorStake
+        // Returns None if nomination not found
+        pub fn rm_nomination(&mut self, validator: AccountId) -> Option<Balance> {
+            let mut amt: Option<Balance> = None;
+            let nominations = self
+                .nominations
+                .0
+                .iter()
+                .filter_map(|x| {
+                    if x.owner == validator {
+                        amt = Some(x.amount);
+                        None
+                    } else {
+                        Some(x.clone())
+                    }
+                })
+                .collect();
+            if let Some(balance) = amt {
+                self.nominations = OrderedSet::from(nominations);
+                // self.total -= balance;
+                self.active_bond -= balance;
+                Some(self.active_bond)
+            } else {
+                None
+            }
+        }
+        // Returns None if nomination not found
+        pub fn inc_nomination(&mut self, validator: AccountId, more: Balance) -> Option<Balance> {
+            for x in &mut self.nominations.0 {
+                if x.owner == validator {
```

Comment:
My understanding is that `self.nominations` is sorted so you can do a binary search here to be more efficient

Reply:
:+1:  Done